pkg/tokens/autofetcher/fetcher.go

```diff
@@ -0,0 +1,273 @@
+package autofetcher
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/status-im/go-wallet-sdk/pkg/tokens/fetcher"
+	"github.com/status-im/go-wallet-sdk/pkg/tokens/parsers"
+	"github.com/status-im/go-wallet-sdk/pkg/tokens/types"
+)
+
+var (
+	ErrStoredListOfTokenListsIsEmpty = errors.New("stored list of token lists is empty")
+	ErrContentStoreNotProvided       = errors.New("content store not provided")
+)
+
+// autofetcher handles the background fetch of token lists (thread-safe for concurrent access).
+type autofetcher struct {
+	mu        sync.Mutex
+	cancel    context.CancelFunc
+	refreshCh chan error
+
+	contentStore ContentStore
+	fetcher      fetcher.Fetcher
+	wg           sync.WaitGroup
+
+	remoteListOfTokenListsFetchDetails types.ListDetails
+	remoteListOfTokenListsParser       parsers.ListOfTokenListsParser
+
+	tokenLists []types.ListDetails
+
+	lastUpdate               time.Time
+	autoRefreshInterval      time.Duration
+	autoRefreshCheckInterval time.Duration // must be <= AutoRefreshInterval
+}
+
+// NewAutofetcherFromRemoteListOfTokenLists creates a new autofetcher from the remote list of token lists.
+// It fetches the remote list of token lists defined by the config and fetches all the token lists from the remote list
+// of token lists and stores them in the content store.
+func NewAutofetcherFromRemoteListOfTokenLists(config ConfigRemoteListOfTokenLists,
+	contentStore ContentStore) (*autofetcher, error) {
+	if err := config.Validate(); err != nil {
+		return nil, err
+	}
+
+	if contentStore == nil {
+		return nil, ErrContentStoreNotProvided
+	}
+
+	return &autofetcher{
+		fetcher:      fetcher.New(fetcher.DefaultConfig()),
+		contentStore: contentStore,
+
+		remoteListOfTokenListsFetchDetails: config.RemoteListOfTokenListsFetchDetails,
+		remoteListOfTokenListsParser:       config.RemoteListOfTokenListsParser,
+
+		lastUpdate:               config.LastUpdate,
+		autoRefreshInterval:      config.AutoRefreshInterval,
+		autoRefreshCheckInterval: config.AutoRefreshCheckInterval,
+	}, nil
+}
+
+// NewAutofetcherFromTokenLists creates a new autofetcher from the token lists.
+// It fetches all the token lists defined by the config and stores them in the content store.
+func NewAutofetcherFromTokenLists(config ConfigTokenLists,
+	contentStore ContentStore) (*autofetcher, error) {
+	if err := config.Validate(); err != nil {
+		return nil, err
+	}
+
+	if contentStore == nil {
+		return nil, ErrContentStoreNotProvided
+	}
+
+	return &autofetcher{
+		fetcher:      fetcher.New(fetcher.DefaultConfig()),
+		contentStore: contentStore,
+
+		tokenLists: config.TokenLists,
+
+		lastUpdate:               config.LastUpdate,
+		autoRefreshInterval:      config.AutoRefreshInterval,
+		autoRefreshCheckInterval: config.AutoRefreshCheckInterval,
+	}, nil
+}
+
+// Start starts the background autofetcher process.
+func (a *autofetcher) Start(ctx context.Context) (refreshCh chan error) {
```

Comment:
We've had several discussions in the Guild times and settled for the approach:
- Long-lived things (services/controllers) have their lifetime managed via a closeCh (created with a Start function call, closed with a Stop function call)
- Short-lived things (fetchers) have their lifetime managed via a ctx passed as argument in their "run" method

Having both a ctx passed here and a Stop function gives you two ways of stopping things that don't even have the exact same effect.